File: app/src/main/java/com/komica/reader/viewmodel/MainViewModel.java
package com.komica.reader.viewmodel;

import android.app.Application;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.MediatorLiveData;

import com.komica.reader.model.Board;
import com.komica.reader.model.BoardCategory;
import com.komica.reader.repository.KomicaRepository;
import com.komica.reader.util.FavoritesManager;

import java.util.ArrayList;
import java.util.List;

public class MainViewModel extends AndroidViewModel {
    private final KomicaRepository repository;
    private final FavoritesManager favoritesManager;
    private final MutableLiveData<List<BoardCategory>> originalCategories = new MutableLiveData<>();
    private final MediatorLiveData<List<BoardCategory>> displayCategories = new MediatorLiveData<>();
    private final MutableLiveData<Boolean> isLoading = new MutableLiveData<>(false);
    private final MutableLiveData<String> errorMessage = new MutableLiveData<>();

    public MainViewModel(@NonNull Application application) {
        super(application);
        repository = KomicaRepository.getInstance(application);
        favoritesManager = FavoritesManager.getInstance(application);
        
        displayCategories.addSource(originalCategories, categories -> updateDisplayCategories());
    }

    public LiveData<List<BoardCategory>> getCategories() {
        return displayCategories;
    }

    public LiveData<Boolean> getIsLoading() {
        return isLoading;
    }

    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }

    public void loadBoards() {
        loadBoards(false);
    }

    public void loadBoards(boolean forceRefresh) {
        isLoading.setValue(true);
        LiveData<List<BoardCategory>> source = repository.fetchBoards(forceRefresh);
        displayCategories.addSource(source, result -> {
            displayCategories.removeSource(source);
            isLoading.setValue(false);
            if (result != null) {
                originalCategories.setValue(result);
            } else {
                errorMessage.setValue("載入板塊失敗");
            }
        });
    }

    public void toggleFavorite(Board board) {
        favoritesManager.toggleFavorite(board.getUrl());
        updateDisplayCategories();
    }

    public void refreshFavorites() {
        updateDisplayCategories();
    }

    private void updateDisplayCategories() {
        List<BoardCategory> original = originalCategories.getValue();
        if (original == null) return;

        List<BoardCategory> updated = new ArrayList<>();
        List<Board> favoriteBoards = new ArrayList<>();

        for (BoardCategory cat : original) {
            for (Board board : cat.getBoards()) {
                if (favoritesManager.isFavorite(board.getUrl())) {
                    boolean alreadyAdded = false;
                    for (Board fb : favoriteBoards) {
                        if (fb.getUrl().equals(board.getUrl())) {
                            alreadyAdded = true;
                            break;
                        }
                    }
                    if (!alreadyAdded) {
                        favoriteBoards.add(board);
                    }
                }
            }
        }

        if (!favoriteBoards.isEmpty()) {
            BoardCategory favCategory = new BoardCategory("我的最愛", favoriteBoards);
            favCategory.setExpanded(true);
            updated.add(favCategory);
        }

        updated.addAll(original);
        displayCategories.setValue(updated);
    }
}

File: app/src/main/java/com/komica/reader/viewmodel/MainViewModel.kt
package com.komica.reader.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.komica.reader.data.FavoritesStore
import com.komica.reader.data.KomicaRepository
import com.komica.reader.model.Board
import com.komica.reader.model.BoardCategory
import com.komica.reader.model.UiState
import com.komica.reader.util.ErrorMessage
import kotlinx.coroutines.launch

class MainViewModel(application: Application) : AndroidViewModel(application) {
    private val repository = KomicaRepository()
    private val favoritesStore = FavoritesStore(application)
    private val originalCategories = mutableListOf<BoardCategory>()
    private var query = ""

    private val _state = MutableLiveData<UiState<List<BoardCategory>>>(UiState.Loading)
    val state: LiveData<UiState<List<BoardCategory>>> = _state

    fun LoadBoards(forceRefresh: Boolean = false) {
        viewModelScope.launch {
            _state.value = UiState.Loading
            runCatching { repository.LoadBoards(forceRefresh) }
                .onSuccess { categories ->
                    originalCategories.clear()
                    originalCategories.addAll(categories)
                    PublishCategories()
                }
                .onFailure { _state.value = UiState.Error(ErrorMessage.Build("載入看板失敗", it), it) }
        }
    }

    fun SetQuery(value: String) {
        query = value.trim()
        PublishCategories()
    }

    fun ToggleFavorite(board: Board) {
        favoritesStore.Toggle(board.url)
        PublishCategories()
    }

    fun IsFavorite(board: Board): Boolean {
        return favoritesStore.IsFavorite(board.url)
    }

    private fun PublishCategories() {
        val filtered = originalCategories.mapNotNull { category ->
            val boards = category.boards.filter { board ->
                query.isBlank() || board.name.contains(query, ignoreCase = true) ||
                    board.url.contains(query, ignoreCase = true) ||
                    board.categoryName.contains(query, ignoreCase = true)
            }
            if (boards.isEmpty()) null else category.copy(boards = boards)
        }

        val favorites = originalCategories
            .flatMap { it.boards }
            .filter { favoritesStore.IsFavorite(it.url) }
            .distinctBy { it.url }

        val result = buildList {
            if (favorites.isNotEmpty()) add(BoardCategory("我的最愛", favorites))
            addAll(filtered)
        }

        _state.value = if (result.isEmpty()) UiState.Empty("目前沒有看板資料") else UiState.Content(result)
    }
}
